Report/sections/3_implementation/scope_and_type_checking.tex

```diff
@@ -0,0 +1,141 @@
+            ? EnvironmentStack.Peek() 
+            : null;
+        SymbolTable newScope = new SymbolTable(parentScope); 
+        
+        parentScope?.Children.Add(newScope); |\label{line:AddingChildToParent}|
+        
+        if (parentScope == null && !TopSymbolTables.ContainsKey(GameObjectIdentifier)) |\label{line:BeginTopScopeCheck}|
+        {
+            TopSymbolTables.Add(GameObjectIdentifier, newScope);
+        } |\label{line:EndTopScopeCheck}|
+
+        EnvironmentStack.Push(newScope);
+        CurrentScope = newScope;
+    }
+
+    protected void CloseScope()
+    {
+        CurrentScope = CurrentScope.Parent;
+    }
+}
+\end{lstlisting}
+
+The method \texttt{OpenScope} is responsible for creating a new scope. Once a new scope is opened, \texttt{OpenScope} assigns its parent to the scope currently on top of the \textbf{EnvironmentStack} if the stack is non-empty. This way, we ensure that there always is access to the parent scopes during scope checking. 
+In addition, the new scope is added as a child of the parent scope such that we can also traverse into child scopes when performing member access later. This can be seen on lines \ref{line:IsStackEmpty} through \ref{line:AddingChildToParent}.
+
+
+As can be seen on line \ref{line:TopSymbolTables}, a dictionary containing the top most scope of all \textbf{GameObjects} is defined. When a new scope is created in \texttt{OpenScope} and it has no parent scope, it is assumed to be the top most scope of the given \textbf{GameObject}.
+If this is the case and the dictionary does not already contain a key with the \textbf{GameObject's} identifier, we add it to the top symbol table dictionary for member access later. This can be seen on lines \ref{line:BeginTopScopeCheck} through \ref{line:EndTopScopeCheck}.
+Afterwards, the new scope is pushed onto the \textbf{EnvironmentStack}. 
+Finally, at the end of the method, the current scope is updated to be the new scope such that the implementing class can use it for type checking. 
+
+As the name implies, the method \texttt{CloseScope} closes the scope by setting the current scope to the parent of the current scope. 
+
+
+The class \textbf{TypeChecker} implements this abstract class. \textbf{TypeChecker} then has access to the methods \texttt{OpenScope} and \texttt{CloseScope}. It also implements the visitor methods that are responsible for traversing the AST to perform both scope and type checking.
+
+As mentioned previously, once the \texttt{PerformSemanticAnalysis(ast)} shown in \snipref{lst:RunMethodSecond} is called, scope checking and type checking begins by visiting the AST. The first step in this visitor sequence is to create a new TypeChecker and call the \texttt{Visit} method for every \textbf{GameObject}, such that the entire AST gets visited to handle scope and type checking - Recall that \textbf{GameObject} represents the root of every subtree in the overall AST. The code for this can be seen in \snipref{lst:TypeCheckEachGO}.
```

Comment:
```suggestion
As mentioned previously, once the \texttt{PerformSemanticAnalysis(ast)} shown in \snipref{lst:RunMethodSecond} is called, scope checking and type checking begins by visiting the AST. The first step in this visitor sequence is to create a new TypeChecker and call the \texttt{Visit} method for every \textbf{GameObject}, such that the entire AST gets visited to handle scope and type checking. The code for this can be seen in \snipref{lst:TypeCheckEachGO}.
```

Reply:
```suggestion
As mentioned previously, once the \texttt{PerformSemanticAnalysis(ast)} shown in \snipref{lst:RunMethodSecond} is called, scope checking and type checking begins by visiting the AST. The first step in this visitor sequence is to create a new TypeChecker and call the \texttt{Visit} method for every \textbf{GameObject}, such that the entire AST gets visited to handle scope and type checking. The reason for visiting each \textbf{GameObject} is that they represents the root of every subtree in the overall AST. The code for this can be seen in \snipref{lst:TypeCheckEachGO}.
```